src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -32,98 +32,46 @@
-    _pos += sz;
-    return true;
-  }
-
-  return false;
-}
-
-void AsyncLogWriter::Buffer::push_flush_token() {
-  bool result = push_back(nullptr, AsyncLogWriter::None, "");
-  assert(result, "fail to enqueue the flush token.");
-}
-
-void AsyncLogWriter::enqueue_locked(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg) {
-  // To save space and streamline execution, we just ignore null message.
-  // client should use "" instead.
-  assert(msg != nullptr, "enqueuing a null message!");
-
-  if (!_buffer->push_back(output, decorations, msg)) {
-    bool p_created;
-    uint32_t* counter = _stats.put_if_absent(output, 0, &p_created);
-    *counter = *counter + 1;
-    return;
-  }
-
-  _data_available = true;
-  _lock.notify();
-}
-
 void AsyncLogWriter::enqueue(LogFileStreamOutput& output, const LogDecorations& decorations, const char* msg) {
-  AsyncLogLocker locker;
-  enqueue_locked(&output, decorations, msg);
+  size_t size = strlen(msg);
+  _circular_buffer.enqueue(msg, size+1, &output, decorations);
 }
 
 // LogMessageBuffer consists of a multiple-part/multiple-line message.
 // The lock here guarantees its integrity.
 void AsyncLogWriter::enqueue(LogFileStreamOutput& output, LogMessageBuffer::Iterator msg_iterator) {
-  AsyncLogLocker locker;
-
-  for (; !msg_iterator.is_at_end(); msg_iterator++) {
-    enqueue_locked(&output, msg_iterator.decorations(), msg_iterator.message());
-  }
+  _circular_buffer.enqueue(output, msg_iterator);
 }
 
-AsyncLogWriter::AsyncLogWriter()
-  : _flush_sem(0), _lock(), _data_available(false),
-    _initialized(false),
-    _stats() {
+AsyncLogWriter::AsyncLogWriter(bool should_stall)
+:
+  _stats_lock(),
+  _stats(),
+  _circular_buffer(_stats, _stats_lock, align_up(AsyncLogBufferSize, os::vm_page_size()), should_stall),
+  _initialized(false) {
 
-  size_t size = AsyncLogBufferSize / 2;
-  _buffer = new Buffer(size);
-  _buffer_staging = new Buffer(size);
-  log_info(logging)("AsyncLogBuffer estimates memory use: " SIZE_FORMAT " bytes", size * 2);
+  log_info(logging)("AsyncLogBuffer estimates memory use: " SIZE_FORMAT " bytes", align_up(AsyncLogBufferSize, os::vm_page_size()));
   if (os::create_thread(this, os::asynclog_thread)) {
     _initialized = true;
   } else {
     log_warning(logging, thread)("AsyncLogging failed to create thread. Falling back to synchronous logging.");
   }
 }
 
-void AsyncLogWriter::write(AsyncLogMap<AnyObj::RESOURCE_AREA>& snapshot) {
+bool AsyncLogWriter::write(AsyncLogMap<AnyObj::RESOURCE_AREA>& snapshot,
+                           char* write_buffer, size_t write_buffer_size) {
   int req = 0;
-  auto it = _buffer_staging->iterator();
-  while (it.hasNext()) {
-    const Message* e = it.next();
+  CircularStringBuffer::Message msg;
+  while (_circular_buffer.has_message()) {
+    CircularStringBuffer::DequeueResult result = _circular_buffer.dequeue(&msg, write_buffer, write_buffer_size);
```

Comment:
Nit: is there any import/using/friend style directive we can use to avoid writing `CircularStringBuffer::DequeueResult` in full each time?

Reply:
Fixed